### PR TITLE
Tests for transaction versioning and interfaces fix

### PR DESCRIPTION
**Pull request by carlpulley-da**
* Tests for transaction versioning and interfaces fix

CHANGELOG_BEGIN

* Ensure SBUBeginExercise has a version that is max(template version, interface version)

CHANGELOG_END

- [x] Add in transaction version tests for interfaces and implements
- [x] Add in transaction version tests for interfaces and coimplements
- [x] ready to merge PR #14765 into https://github.com/digital-asset/daml/pull/14709

### CODE REVIEW

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TransactionVersionTest.scala [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+
+  "interface and transaction versioning" - {
+
+    "version testing assumptions" in {
+      oldVersion should be < newVersion
+      Set(
+        templatePkg.languageVersion,
+        interfacesPkg.languageVersion,
+        implementsPkg.languageVersion,
+//        coImplementsPkg.languageVersion,
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TransactionVersionTest.scala:32]: Code graded out for preliminary code review

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TransactionVersionTest.scala [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+      ) shouldBe Set(commonVersion)
+    }
+
+    "template version > interface version" in {
+      val oldPkg1 = templatePkg.copy(languageVersion = oldVersion)
+      val oldPkg2 = interfacesPkg.copy(languageVersion = oldVersion)
+      val newPkg1 = implementsPkg.copy(languageVersion = newVersion)
+//      val newPkg2 = coImplementsPkg.copy(languageVersion = newVersion)
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TransactionVersionTest.scala:40]: Code graded out for preliminary code review

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TransactionVersionTest.scala [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+      val pkgs = SpeedyTestLib.typeAndCompile(
+        Map(
+          templatePkgId -> oldPkg1,
+          interfacesPkgId -> oldPkg2,
+          implementsPkgId -> newPkg1,
+//          coImplementsPkgId -> newPkg2,
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TransactionVersionTest.scala:46]: Code graded out for preliminary code review

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TransactionVersionTest.scala [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+      )
+
+      inside(result) { case Right(transaction) =>
+        transaction.version shouldBe TransactionVersion.assignNodeVersion(newVersion)
+      }
+    }
+
+    "template version < interface version" in {
+      val oldPkg1 = implementsPkg.copy(languageVersion = oldVersion)
+//      val oldPkg2 = coImplementsPkg.copy(languageVersion = oldVersion)
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TransactionVersionTest.scala:66]: Code graded out for preliminary code review

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TransactionVersionTest.scala [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+      val newPkg1 = templatePkg.copy(languageVersion = oldVersion)
+      val newPkg2 = interfacesPkg.copy(languageVersion = newVersion)
+      val pkgs = SpeedyTestLib.typeAndCompile(
+        Map(
+          templatePkgId -> newPkg1,
+          interfacesPkgId -> newPkg2,
+          implementsPkgId -> oldPkg1,
+//          coImplementsPkgId -> oldPkg2,
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TransactionVersionTest.scala:74]: Code graded out for preliminary code review

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TransactionVersionTest.scala [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+      }
+    }
+
+    "template version == interface version" in {
+      val pkgs = SpeedyTestLib.typeAndCompile(
+        Map(
+          templatePkgId -> templatePkg,
+          interfacesPkgId -> interfacesPkg,
+          implementsPkgId -> implementsPkg,
+//          coImplementsPkgId -> coImplementsPkg,
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TransactionVersionTest.scala:98]: Code greyed out for preliminary code review

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TransactionVersionTest.scala [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+            implements 'interfaces-pkg':InterfacesMod:Interface2 {
+              view = ();
+              method getPerson = 'implements-pkg':ImplementsMod:TemplateImplements12 {person} this;
+              method getLabel = "template-implements-1-2";
+            };
+          };
+        }
+      """
+
+//  val (coImplementsPkgId, coImplementsPkg) =
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TransactionVersionTest.scala:212]: Code greyed out for preliminary code review

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TransactionVersionTest.scala [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+    val speedyObservers = SExpr.SEValue(SValue.SList(FrontStack.Empty))
+    val machine =
+      Speedy.Machine.fromUpdateSExpr(
+        pkgs,
+        transactionSeed = crypto.Hash.hashPrivateKey("TransactionVersionTest"),
+        updateSE = SEMakeClo(
+          Array(),
+          1,
+          SExpr.SELet1General(
+            SBFetchAny(speedyContractId, SEValue.None),
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TransactionVersionTest.scala:280]: `SBFecthAny` is used to ensure the contract finds its way into the ledger cage (ow `SBUBeginExercise` will exception).

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TransactionVersionTest.scala [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+            SExpr.SEScopeExercise(
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TransactionVersionTest.scala:281]: `SEScopeExercise` is used to ensure that a root context is present at the end of command processing (ow Speedy will exception).

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TransactionVersionTest.scala [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+              observers (Nil @Party)
+              to upure @Unit ();
+          };
+        }
+       """
+  val (interfacesPkgId, interfacesPkg) =
+    PackageId.assertFromString("interfaces-pkg") -> p"""
+         module InterfacesMod {
+           interface (this: Interface1) = {
+             viewtype Unit;
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TransactionVersionTest.scala:157]: please use a record type, as we will soon check views are records.
see #14746 for examples

**Review by remyhaemmerle-da — APPROVED** [review]
Thanks.

**Pull request merged**